pkg/registrar/registrar.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registrar
+
+import (
+	"fmt"
+	"sync"
+)
+
+// TODO(random-liu): Add unit test.
+
+// Registrar stores name to key, and key to name indexes.
+// Names must be unique.
+// Registrar is safe for concurrent access.
+type Registrar struct {
+	sync.Mutex
```

Comment:
nit: Just a personal preference so feel free to ignore. I'd declare `lock sync.Mutex` so that `Lock()/Unlock()` won't be exported and accessed by accident.